File: GFG POTD/011-October-2024/016-two-swaps.cpp
// https://www.geeksforgeeks.org/problems/two-swaps--155623/1




//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;
// } Driver Code Ends

class Solution {
  public:
    bool checkSorted(vector<int> &arr) {
        int n = arr.size();
        
        int cntSwaps = 0;
        for (int i = 0; i < n; i ++) {
            if (arr[i] - 1 != i) {
                swap(arr[i], arr[arr[i] - 1]);
                i --;
                cntSwaps ++;
            }
        }
        
        return true ? cntSwaps == 0 || cntSwaps == 2 : false;
    }
};


//{ Driver Code Starts.
int main() {

    int t;
    cin >> t;
    cin.ignore();
    while (t--) {
        string input;
        getline(cin, input);
        stringstream ss(input);
        int num;
        vector<int> arr;
        while (ss >> num)
            arr.push_back(num);

        Solution ob;
        bool ans = ob.checkSorted(arr);
        if (ans)
            cout << "true" << endl;
        else
            cout << "false" << endl;
    }
}
// } Driver Code Ends

File: GFG POTD/011-October-2024/016-two-swaps.java
// https://www.geeksforgeeks.org/problems/two-swaps--155623/1




//{ Driver Code Starts
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int t = Integer.parseInt(scanner.nextLine()); // Read number of test cases

        // Loop through each test case
        while (t-- > 0) {
            String input = scanner.nextLine();
            String[] inputArr = input.split(" ");
            List<Integer> arr = new ArrayList<>();

            // Convert input to list of integers
            for (String str : inputArr) {
                arr.add(Integer.parseInt(str));
            }

            Solution ob = new Solution();
            boolean ans = ob.checkSorted(arr);

            // Output result
            if (ans)
                System.out.println("true");
            else
                System.out.println("false");
        }

        scanner.close();
    }
}

// } Driver Code Ends
class Solution {
    public boolean checkSorted(List<Integer> arr) {
        int n = arr.size();
        int cntSwaps = 0;
        for (int i = 0; i < n; i ++) {
            if (arr.get(i) - 1 != i) {
                Collections.swap(arr, i, arr.get(i) - 1);
                cntSwaps ++;
                i --;
            }
        }
        return true ? cntSwaps == 0 || cntSwaps == 2 : false;
    }
}
